misoc/cores/duc.py

```diff
@@ -205,29 +205,21 @@ def __init__(self, width, constants):
 
         ###
 
+        self.latency = 2 if n > 8 else 0
+
         # TODO: improve MCM
-        assert n <= 9
+        assert n <= 16
         assert range(n) == constants
 
-        ctx = self.comb
-        if n > 0:
-            ctx += o[0].eq(0)
-        if n > 1:
-            ctx += o[1].eq(i)
-        if n > 2:
-            ctx += o[2].eq(i << 1)
-        if n > 3:
-            ctx += o[3].eq(i + (i << 1))
-        if n > 4:
-            ctx += o[4].eq(i << 2)
-        if n > 5:
-            ctx += o[5].eq(i + (i << 2))
-        if n > 6:
-            ctx += o[6].eq(o[3] << 1)
-        if n > 7:
-            ctx += o[7].eq((i << 3) - i)
-        if n > 8:
-            ctx += o[8].eq(i << 3)
+        ctx = self.sync if n > 8 else self.comb
```

Comment:
Don't do that. It's architecture dependent and unexpected. Just add a register layer after in downstream code if you need it.